lib/src/kvs/ds.rs

```diff
@@ -524,62 +559,186 @@ impl Datastore {
-		let val = self.remove_archived(&mut tx, filtered).await;
-		let resolve_err = match val {
-			Ok(_) => tx.commit().await,
-			Err(e) => {
-				error!("Error bootstrapping sweep phase: {:?}", e);
-				match tx.cancel().await {
-					Ok(_) => Err(e),
-					Err(e) => {
-						// We have a nested error
-						Err(Error::Tx(format!("Error bootstrapping sweep phase: {:?} and error cancelling transaction: {:?}", e, e)))
+		// The transaction request pair is used for receiving requests for new transactions
+		// it contains one-shot senders so that the transaction can be sent and forgotten
+		let (tx_req_send, mut tx_req_recv): (
+			mpsc::Sender<oneshot::Sender<SendTransaction>>,
+			mpsc::Receiver<oneshot::Sender<SendTransaction>>,
+		) = mpsc::channel(BOOTSTRAP_BATCH_SIZE);
+
+		// In several new transactions, scan all removed node live queries
+		let (scan_send, scan_recv): (
+			mpsc::Sender<BootstrapOperationResult>,
+			mpsc::Receiver<BootstrapOperationResult>,
+		) = mpsc::channel(BOOTSTRAP_BATCH_SIZE);
+
+		let scan_task = bootstrap::scan_node_live_queries(
+			tx_req_send.clone(),
+			dead_nodes,
+			scan_send,
+			BOOTSTRAP_BATCH_SIZE as u32,
+		);
+
+		// Wasm doesnt have multithreading
+		#[cfg(not(target_arch = "wasm32"))]
+		let mut scan_task = tokio::spawn(scan_task);
+
+		// In several new transactions, archive removed node live queries
+		let (archive_send, archive_recv): (
+			mpsc::Sender<BootstrapOperationResult>,
+			mpsc::Receiver<BootstrapOperationResult>,
+		) = mpsc::channel(BOOTSTRAP_BATCH_SIZE);
+		trace!("Spawned archive task");
+		let archive_task = bootstrap::archive_live_queries(
+			tx_req_send.clone(),
+			self.id,
+			scan_recv,
+			archive_send,
+			BOOTSTRAP_BATCH_SIZE,
+			&BOOTSTRAP_BATCH_LATENCY,
+		);
+
+		// Wasm doesnt have multithreading
+		#[cfg(not(target_arch = "wasm32"))]
+		let mut archive_task = tokio::spawn(archive_task);
+
+		// In several new transactions, delete archived node live queries
+		let (delete_send, delete_recv): (
+			mpsc::Sender<BootstrapOperationResult>,
+			mpsc::Receiver<BootstrapOperationResult>,
+		) = mpsc::channel(BOOTSTRAP_BATCH_SIZE);
+		trace!("Spawned delete live query task");
+		let delete_task = bootstrap::delete_live_queries(
+			tx_req_send,
+			archive_recv,
+			delete_send,
+			BOOTSTRAP_BATCH_SIZE,
+		);
+		// Wasm doesnt have multithreading
+		#[cfg(not(target_arch = "wasm32"))]
+		let mut delete_task = tokio::spawn(delete_task);
+
+		// We then need to collect and log the errors
+		// It's also important to consume from the channel otherwise things will block
+		trace!("Spawned delete handler task");
+		let delete_handler_task = create_delete_handler_future(delete_recv);
+
+		// Wasm doesnt have multithreading
+		#[cfg(not(target_arch = "wasm32"))]
+		let mut delete_handler_task = tokio::spawn(delete_handler_task);
+		#[cfg(target_arch = "wasm32")]
+		let mut delete_handler_task = Pin::new(delete_handler_task);
```

Comment:
This shouldn't be necessary. You can spawn tasks on Wasm using `wasm_bindgen_futures::spawn_local`. A "dummy waker" shouldn't be necessary.

Reply:
Yes, but they are blocking and dont have a return type. This code cannot be blocking. As for the waker, that is required for the context I think. Would be very excited not to have a waker though

Reply:
Futures spawned with spawn_local should not block.
This code:
```rust
use gloo_timers::future::TimeoutFuture;
use wasm_bindgen::prelude::*;
use wasm_bindgen_futures::spawn_local;
use wasm_timer::Instant;

#[wasm_bindgen]
extern "C" {
    #[wasm_bindgen(js_namespace = console)]
    fn log(s: &str);
}

#[wasm_bindgen]
pub fn test() {
    log("STARTING");
    let before = Instant::now();
    spawn_local(async move {
        TimeoutFuture::new(1000).await;
        log(&format!("ELAPSED A: {}", before.elapsed().as_secs_f32()));
    });
    spawn_local(async move {
        TimeoutFuture::new(1000).await;
        log(&format!("ELAPSED B: {}", before.elapsed().as_secs_f32()));
    });
}
```
prints
```
STARTING
ELAPSED A: 1.001
ELAPSED B: 1.001
```
In the browser, so neither future is blocking the other.

Reply:
And you could fix the return type with a channel.

Reply:
Yes, I'm not sure where @phughk got the idea that it blocks. Even [the documentation](https://docs.rs/wasm-bindgen-futures/0.4.40/wasm_bindgen_futures/fn.spawn_local.html) explicitly says

> The `future` must be `'static` because it will be scheduled to run in the background